s3-migrate/s3-copy.py

```diff
@@ -0,0 +1,228 @@
+            try:
+                self.client.put_object(
+                    Body=open(srcBucket.tmpMapFile, "rb"), Bucket=self.bucket, Key=file
+                )
+            except Exception as e:
+                logging.error(e)
+            else:
+                break
+        else:
+            logging.info(
+                "L'upload du fichier %s sur le bucket %s a échoué" % (file, self.bucket)
+            )
+
+    def deleteFile(self, file):
+        logging.info("Suppression du fichier %s et de ses versions" % file)
+        for attempt in range(10):
+            try:
+                objects = []
+                listVersion = self.getVersionsOfDoc(file)
+                for v in listVersion:
+                    objects.append({"VersionId": v["VersionId"], "Key": v["Key"]})
+                if objects:
+                    self.client.delete_objects(
+                        Bucket=self.bucket, Delete={"Objects": objects}
+                    )
+                else:
+                    logging.info(
+                        "Le fichier n'existe pas sur le bucket %s" % self.bucket
+                    )
+            except Exception as e:
+                logging.error(e)
+            else:
+                break
+        else:
+            logging.info("La suppression du fichier %s a échoué" % file)
+
+
+src = Bucket(
+    os.environ.get("SRC_ENDPOINT_URL", "http://localhost:9000"),
+    os.environ.get("SRC_ACCESS_KEY", "changeme"),
+    os.environ.get("SRC_SECRET_KEY", "changeme"),
+    os.environ.get("SRC_BUCKET_NAME", "test1"),
+    os.environ.get("PAGINATION_SIZE", 1000),
+)
+dst = Bucket(
+    os.environ.get("DST_ENDPOINT_URL", "http://localhost:9000"),
+    os.environ.get("DST_ACCESS_KEY", "changeme"),
+    os.environ.get("DST_SECRET_KEY", "changeme"),
+    os.environ.get("DST_BUCKET_NAME", "test1"),
+    os.environ.get("PAGINATION_SIZE", 1000),
+)
+
+logging.info("Initialisation")
+for file in src.getDocumentsList():
+    logging.info("Début de la boucle sur la liste des fichiers")
+    changeDetected = True
+    updateMeta = True
+    if "meta.json" not in file:
+        srcMeta = src.searchMeta(file)
+        dstMeta = dst.searchMeta(file)
+        if srcMeta:
+            if dstMeta:
+                logging.info(
+                    "meta.json trouvé sur la  cible, vérification nouvelle version"
+                )
+                if (
+                    srcMeta["size"] == dstMeta["size"]
+                    and srcMeta["lastModified"] == dstMeta["lastModified"]
+                ):
+                    logging.info("Pas de nouvelle version détectée")
+                    changeDetected = False
+                else:
+                    logging.info(
+                        "Nouvelle version détecté pour %s, suppression sur la cible"
+                        % file
+                    )
+                    dst.deleteFile(file)
+                    dst.deleteFile(srcMeta["uri"])
```

Comment:
I would remove these two files only when the source file is removed. Also synchronize the versions of the destination with the versions of the source.

I have ideas in mind, which imply using the `meta.json`'s `metadata` or `lastModified` fields to map the versions of the source and of the destination.

The algorithm should be quite simple, as missing versions in dst (so versions we should upload) are necessarily newer than the existing versions. Something to illustrate the idea below:
![idea for synchronizing versions](https://github.com/user-attachments/assets/a75db8c6-6715-4a8d-bcbe-6ab9ffe3173f)


Something like:
```python
srcVersions = src.readVersionsFromMetaJson(file)
dstVersions = dst.readVersionsFromMetaJson(file)
srcVersionsByLastModified = {item['lastModified']: item for item in srcVersions} # Use lastModified as a common key between the source and the destination
dstVersionsByLastModified = {item['lastModified']: item for item in dstVersions}

versionsLastModifiedToDelete = dstVersionsByLastModified.keys() - srcVersionsByLastModified.keys()
versionsToDeleteInDst = [versionsLastModifiedToDelete[item['lastModified']]['versionId'] for item in versionsLastModifiedToDelete] # Retrieve each versionId on the destination for each `lastModified` values.
dst.deleteVersions(versionsToDeleteInDst)

versionsLastModifiedToUpload = srcVersionsByLastModified.keys() - dstVersionsByLastModified.keys()
versionsToUploadToDst = [versionsLastModifiedToUpload[item['lastModified']]['versionId'] for item in versionsLastModifiedToUpload] # Retrieve each versionId on the destination for each `lastModified` values.
dst.uploadVersions(versionsToUploadToDst)
```

Reply:
As an alternative, maybe less hacky'ish, we may introduce a file which would map the source `versionId`s with the destination `versionId`s for each file.